pkg/issuer/vault/issue.go

```diff
@@ -0,0 +1,200 @@
+	if k8sErrors.IsNotFound(err) || errors.IsInvalidData(err) {
+		signeeKey, err = pki.GenerateRSAPrivateKey(keyBitSize)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error generating private key: %s", err.Error())
+		}
+	}
+
+	if err != nil {
+		return nil, nil, fmt.Errorf("error getting certificate private key: %s", err.Error())
+	}
+
+	commonName := crt.Spec.CommonName
+	altNames := crt.Spec.DNSNames
+	if len(commonName) == 0 && len(altNames) == 0 {
+		return nil, nil, fmt.Errorf("no domains specified on certificate")
+	}
+
+	crtPem, err := v.signCertificate(crt, signeeKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return pki.EncodePKCS1PrivateKey(signeeKey), crtPem, nil
+}
+
+// signCertificate returns a signed x509.Certificate object for the given
+// *v1alpha1.Certificate crt.
+func (v *Vault) signCertificate(crt *v1alpha1.Certificate, key *rsa.PrivateKey) ([]byte, error) {
+	commonName, err := pki.CommonNameForCertificate(crt)
+	if err != nil {
+		return nil, fmt.Errorf("error extracting Common Name from certificate: %s", err.Error())
+	}
+
+	altNames, err := pki.DNSNamesForCertificate(crt)
+	if err != nil {
+		return nil, fmt.Errorf("error extracting DNS Names from certificate: %s", err.Error())
+	}
+
+	if len(commonName) == 0 && len(altNames) > 0 {
+		commonName = altNames[0]
+	}
+
+	template := pki.GenerateCSR(commonName, altNames...)
+
+	derBytes, err := x509.CreateCertificateRequest(rand.Reader, template, key)
+	if err != nil {
+		return nil, fmt.Errorf("error creating x509 certificate: %s", err.Error())
+	}
+
+	pemRequestBuf := &bytes.Buffer{}
+	err = pem.Encode(pemRequestBuf, &pem.Block{Type: "CERTIFICATE REQUEST", Bytes: derBytes})
+	if err != nil {
+		return nil, fmt.Errorf("error encoding certificate request: %s", err.Error())
+	}
+
+	return v.requestVaultCert(commonName, altNames, pemRequestBuf.String())
+}
+
+func (v *Vault) requestVaultCert(commonName string, altNames []string, csr string) ([]byte, error) {
+	client, err := vault.NewClient(nil)
+	if err != nil {
+		return nil, fmt.Errorf("error initializing Vault client: %s", err.Error())
+	}
+
+	client.SetAddress(v.issuer.GetSpec().Vault.Server)
+
+	tokenRef := v.issuer.GetSpec().Vault.Auth.TokenSecretRef
+	token, err := vaultTokenRef(v.secretsLister, v.issuerResourcesNamespace, tokenRef.Name, "token")
+	if err != nil {
+		return nil, fmt.Errorf("error reading Vault token from secret %s/%s: %s", v.issuerResourcesNamespace, tokenRef.Name, err.Error())
+	}
+
+	client.SetToken(token)
+
+	glog.V(4).Infof("Vault certificate request for commonName %s altNames: %q", commonName, altNames)
+
+	parameters := map[string]string{
+		"common_name": commonName,
+		"alt_names":   strings.Join(altNames, ","),
+		"ttl":         time.Duration(certificateDuration).String(),
+		"csr":         csr,
+		"exclude_cn_from_sans": "true",
+	}
+
+	url := fmt.Sprintf("/v1%s", v.issuer.GetSpec().Vault.Path)
```

Comment:
I think `v.issuer.GetSpec().Vault.Path` should not be required to start with a `/`, as vault itself doesn't use this afaik.

If it *does* start with a `/`, we could also automatically strip it, thus allowing both. But this `fmt.Sprintf` should be using `/v1/%s` imo.

Reply:
I will change it for a path.Join so it will work in both cases.